05/app.js

```diff
@@ -0,0 +1,56 @@
+function Student(name, surname) {
+  this.name = name;
+  this.surname = surname;
+  this.grade = {};
+}
+Student.prototype.addGrade = function (subject, mark) {
+  if (!this.grade.hasOwnProperty(subject)) {
+    this.grade[subject] = [];
+  }
+  this.grade[subject].push(mark);
+};
+Student.prototype.calculateAverage = function (grades) {
+  const sum = grades.reduce(
+    (accumulator, currentValue) => accumulator + currentValue,
+    0
+  );
+  const avg = sum / grades.length;
+  return avg;
+};
+Student.prototype.getAverageGrade = function (subject) {
+  if (subject === undefined) {
+    let avgSum = 0;
+    let count = 0;
+    for (const key in this.grade) {
+      const grades = this.grade[key];
+      const avg = this.calculateAverage(grades);
```

Comment:
W JS (a w zasadzie w większości języków programowania) dane są zapisywane w formacie dwójkowym to powoduje problemy z dokładnością przy operacjach na liczbach zmiennoprzecinkowych (sprawdź 0.1 + 0.2). Dlatego lepiej jest wykonać sumę dla wszystkich (mamy liczby całkowite), a dopiero potem dzielić - przy mniejszym zbiorze danych i zaokrągleniu to nie problem, ale jeśli dokładność jest istotna to już może być :)